src/IoTeX/Signer.cpp

```diff
@@ -44,41 +44,85 @@ Data Signer::hash() const {
     return Hash::keccak256(action.SerializeAsString());
 }
 
-static Data encodeStaking(const Proto::Staking& staking) {
-    Data encoded;
-    if (staking.has_stake()) {
-        auto& stake = staking.stake();
-        stakingStake(TW::data(stake.candidate()), stake.duration(), stake.nondecay(), TW::data(stake.data()), encoded);
-    } else if (staking.has_unstake()) {
-        auto& unstake = staking.unstake();
-        stakingUnstake(unstake.piggy_index(), TW::data(unstake.data()), encoded);
-    } else if (staking.has_withdraw()) {
-        auto& withdraw = staking.withdraw();
-        stakingWithdraw(withdraw.piggy_index(), TW::data(withdraw.data()), encoded);
-    } else if (staking.has_movestake()) {
-        auto& move = staking.movestake();
-        stakingMoveStake(move.piggy_index(), TW::data(move.candidate()), TW::data(move.data()), encoded);
-    } else if (staking.has_addstake()) {
-        auto& add = staking.addstake();
-        stakingAddStake(add.piggy_index(), TW::data(add.data()), encoded);
-    }
-    return encoded;
-}
 void Signer::toActionCore() {
-    if (input.has_staking()) {
-        action.set_version(input.version());
-        action.set_nonce(input.nonce());
-        action.set_gaslimit(input.gaslimit());
-        action.set_gasprice(input.gasprice());
-        auto& staking = input.staking();
-        auto encoded = encodeStaking(staking);
-        auto& execution = *action.mutable_execution();
-        execution.set_amount(staking.amount());
-        execution.set_contract(staking.contract());
-        execution.set_data(encoded.data(), encoded.size());
-    } else {
-        // ActionCore is almost same as SigningInput, missing field privateKey = 5;
-        action.ParseFromString(input.SerializeAsString());
-        action.DiscardUnknownFields();
+    action.set_version(input.version());
```

Comment:
with updated ActionCore and SigningInput (see comment below)
we don't need to assert the type of input for staking
can directly use the else {} part of old code